File: src/com/leftyist/kingdomsim/utils/XMLParser.java
package com.leftyist.kingdomsim.utils;

import org.w3c.dom.Document;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;

import java.io.File;
import java.util.ArrayList;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;

import javafx.util.Pair;

public class XMLParser
{
      public XMLParser() {
      }

      public static Document openFile(String fileName) {
            try {
                  //init
                  File file = new File(fileName);
                  DocumentBuilderFactory documentBuilderFactory = DocumentBuilderFactory.newInstance();
                  documentBuilderFactory.setIgnoringElementContentWhitespace(true);
                  DocumentBuilder documentBuilder = documentBuilderFactory.newDocumentBuilder();
                  return(documentBuilder.parse(file));
            } catch (Exception e) {
                  return null;
            }
      }

      private static String[] getNodeValues(NodeList nodeList) {
            String[] ret = new String[nodeList.getLength()];
            for(int i = 0; i < nodeList.getLength(); i++)
                  ret[i] = nodeList.item(i).getTextContent();

            return ret;
      }

      private static ArrayList<Pair<String, String>> getNodeSubAttributes(NodeList node) {
            ArrayList<Pair<String,String>> values = new ArrayList<>();

            for(int i = 0; i < node.getLength(); i++) {
                  String key = node.item(i).getNodeName();

                  if(key.equals("#text"))
                        continue;

                  Node n = node.item(i);

                  values.add(new Pair<>(n.getNodeName(), n.getTextContent()));
            }
            return values;
      }

      public static String findValueByTag(NodeList nodes, String tagName) {

            for (int i = 0; i < nodes.getLength(); i++) {
                  String nodeName = nodes.item(i).getNodeName();
                  if (nodeName.equals(tagName))
                        return nodes.item(i).getTextContent();
            }
            return null;
      }

      public static NodeList findChild(NodeList nodes, String childName) {
            if (nodes == null)
                  return null;

            for(int i = 0; i < nodes.getLength(); i++) {
                  if(nodes.item(i).getNodeName().equals(childName))
                        return nodes.item(i).getChildNodes();
            }
            return null;
      }
}

File: src/com/leftyist/kingdomsim/utils/XMLParser.kt
package com.leftyist.kingdomsim.utils

import org.w3c.dom.Document
import org.w3c.dom.Node
import org.w3c.dom.NodeList
import java.io.File
import javax.xml.parsers.DocumentBuilderFactory
import javax.xml.transform.TransformerFactory
import javax.xml.transform.dom.DOMSource
import javax.xml.transform.stream.StreamResult

fun openFile(fileName: String): Document {
      try {
            val file = File(fileName)
            val documentBuilderFactory = DocumentBuilderFactory.newInstance()
            documentBuilderFactory.isIgnoringElementContentWhitespace = true
            val documentBuilder = documentBuilderFactory.newDocumentBuilder()
            return documentBuilder.parse(file)
      } catch (e: Exception) {
            throw IllegalArgumentException("File not found")
      }
}

fun NodeList.findChild(childName: String): Node? {
      if (this == null)
            return null

      for (i in 0..this.length - 1) {
            if (this.item(i).nodeName == childName)
                  return this.item(i)
      }
      return null
}

fun Node.findChild(childName: String): Node? {
      return this.childNodes.findChild(childName)
}

fun getNodeSubValues(node: Node): ArrayList<Pair<String, String>> {
      val list = ArrayList<Pair<String, String>>()

      val nodes = node.childNodes

      for (i in 0..nodes.length - 1) {
            val subnode = nodes.item(i)

            if (subnode.nodeName == "#text")
                  continue
            else
                  list.add(Pair(subnode.nodeName, getValue(subnode)))

      }
      return list
}

fun findValueInNode(node: Node, value: String): String {
      val nodes = node.childNodes
      for(i in 0..nodes.length) {
            val subnode = nodes.item(i)
            if(subnode.nodeName == value)
                  return getValue(node)
      }

      return ""
}

fun initFromTemplate(fileNameAndPath: String, templatePath: String) {
      val template = File(templatePath)
      if(!template.exists())
            throw IllegalArgumentException("Template not found.")

      val destFile = File(fileNameAndPath)
      if(destFile.exists()) {
            throw IllegalArgumentException("File already exists.")
      }

      template.copyTo(destFile, false)
}

fun saveFile(doc: Document, path: String) {
      val transformerFactory = TransformerFactory.newInstance()
      val transformer = transformerFactory.newTransformer()
      val source = DOMSource(doc)
      val result = StreamResult(File(path))
      transformer.transform(source, result)
}

private fun getValue(node: Node): String {
      if(node.attributes.getNamedItem("type") != null) {
            val parentName = node.attributes.getNamedItem("parent").nodeValue
            val root = getRootNode(node)

            val parentNode = root.findChild(parentName)
            if(parentNode == null)
                  return "LOOKUP ERROR: WRONG PARENT NAME"

            val subNode = parentNode.findChild(node.textContent)
            if(subNode == null)
                  return "LOOKUP ERROR: WRONG VALUE"

            val valNode = subNode.findChild(node.nodeName)
            if(valNode == null)
                  return "LOOKUP ERROR: WRONG ELEMENT"

            return getValue(valNode)


      } else {
            return node.textContent
      }
}

private fun getRootNode(node: Node): Node {
      var root = node;
      while(root.parentNode != null)
            root = root.parentNode

      root = root.childNodes.item(0)

      return root
}